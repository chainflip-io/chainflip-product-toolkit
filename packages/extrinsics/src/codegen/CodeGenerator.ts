import { capitalize } from '@chainflip/utils/string';
import BaseCodeGenerator, { CodegenResult, Code, Identifier } from '@/chainspec/BaseCodeGenerator';
import {
  ArrayType,
  EnumType,
  MapType,
  OptionType,
  PrimitiveType,
  RangeType,
  ResolvedType,
  StructType,
  TupleType,
} from '@/chainspec/BaseParser';
import assert from 'assert';

class Struct extends Code {
  as: 'tuple' | 'record' = 'record';
  declarationType: 'type' | 'const' = 'type';

  constructor(private readonly fields: (readonly [string, CodegenResult])[]) {
    super(
      '',
      fields.map(([, type]) => type),
    );
  }

  asTuple() {
    this.as = 'tuple';
    return this;
  }

  override toString() {
    let open: string;
    let close: string;
    let transform = (name: string) => name;

    if (this.as === 'tuple') {
      open = '[';
      close = ']';
    } else {
      open = '{';
      close = '}';
      transform = (name) => name.replace(/[A-Z]/g, (m) => `_${m.toLowerCase()}`);
    }

    return `${open}\n${this.fields.map(([name, type]) => `${transform(name)}: ${type.toString()}`).join(',\n')}\n${close}`;
  }

  [Symbol.toStringTag]() {
    return 'Struct';
  }
}

export default class CodeGenerator extends BaseCodeGenerator {
  private getKnownIdentifier(name: string): string {
    switch (name) {
      case 'cfPrimitivesChainsAssetsAnyAsset':
        return 'ChainflipAsset';
      case 'cfChainsAddressEncodedAddress':
        return 'EncodedAddress';
      case 'cfChainsAddressForeignChainAddress':
        return 'ForeignChainAddress';
      case 'cfChainsBtcScriptPubkey':
        return 'BitcoinScriptPubKey';
      default:
        return capitalize(name);
    }
  }

  protected override generatePrimitive(def: PrimitiveType): CodegenResult {
    switch (def.name) {
      case 'u8':
      case 'u16':
      case 'u32':
        return new Code('number').asType();
      case 'u128':
      case 'U256':
      case 'AccountId32':
        return new Code('`0x${string}`').asType();
      case 'Bytes':
        return new Code('Uint8Array | `0x${string}`').asType();
      case 'H160':
        return new Code('Uint8Array').asType();
      default:
        throw new Error(`Method not implemented for primitive type: ${def.name}`);
    }
  }

  protected override generateEnum(def: EnumType): CodegenResult {
    const identifier = this.getKnownIdentifier(def.name);
    if (this.registry.types.has(identifier)) return new Identifier(identifier).asType();

    const isSimpleEnum = def.values.every(
      (value) => value.value.type === 'primitive' && value.value.name === 'null',
    );

    let members: CodegenResult[];

    if (isSimpleEnum) {
      members = def.values
        .filter((value) => !value.name.startsWith('__Unused'))
        .map((value) => new Code(`'${value.name}'`).asType());
    } else {
      members = def.values.map((value) => {
        const result = this.generateResolvedType(value.value);
        return new Code(`Record<'${value.name}', ${result.toString()}>`, [result]).asType();
      });
    }

    const code = new Code(members.join(' | '), members).asType();
    this.registry.types.set(identifier, code);
    return new Identifier(identifier).asType();
  }

  protected override generateStruct(def: StructType): CodegenResult {
    const fields = Object.entries(def.fields).map(([name, type]) => {
      const result = this.generateResolvedType(type);
      return [name, result] as const;
    });
    return new Struct(fields);
  }

  protected override generateArray(def: ArrayType): CodegenResult {
    if (def.value.type === 'primitive' && def.value.name === 'u8') {
      return new Code('Uint8Array').asType();
    }

    const type = this.generateResolvedType(def.value);

    return new Code(`${type.toString()}[]`, [type]);
  }

  protected override generateTuple(_def: TupleType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateRange(_def: RangeType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateOption(def: OptionType): CodegenResult {
    const type = this.generateResolvedType(def.value);
    return new Code(`${type.toString()} | null`, [type]);
  }

  protected override generateMap(_def: MapType): CodegenResult {
    throw new Error('Method not implemented.');
  }

  protected override generateItem(itemName: string, def: ResolvedType): CodegenResult {
    const type = this.generateResolvedType(def);
    assert(def.type === 'struct', `Expected struct, got ${def.type}`);
    assert(type instanceof Struct, `Expected code, got ${type[Symbol.toStringTag]()}`);
    return type.asTuple();
  }

  protected override getParserName(palletName: string, itemName: string): string {
    return `${palletName}.${itemName}`;
  }

  protected override getName(palletName: string, itemName: string): string {
    return `${capitalize(palletName)}${capitalize(itemName).replace(/_(\w)/g, (m, c: string) => c.toUpperCase())}`;
  }

  protected override getFileName(palletName: string, itemName: string): string {
    return itemName.replace(/_(\w)/g, (m, c: string) => c.toUpperCase());
  }
}
