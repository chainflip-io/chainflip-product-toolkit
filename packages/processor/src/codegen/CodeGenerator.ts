import BaseCodeGenerator, { CodegenResult, Code, Identifier } from '@/chainspec/BaseCodeGenerator';
import {
  isPrimitiveType,
  type ArrayType,
  type EnumType,
  type MapType,
  type OptionType,
  type PrimitiveType,
  type RangeType,
  type ResolvedType,
  type StructType,
  type TupleType,
} from './Parser';

const nameToIdentifier = (name: string): string =>
  name
    .replace(/(?:::|_)(.)/g, (_, c: string) => c.toUpperCase())
    .replace(/^./, (c) => c.toLowerCase());

const isNull = (type: ResolvedType) => type.type === 'primitive' && type.name === 'null';

const hexString = new Code(
  "z.string().refine((v): v is `0x${string}` => /^0x[\\da-f]*$/i.test(v), { message: 'Invalid hex string' })",
);
const numericString = new Code(
  "z.string().refine((v) => /^\\d+$/.test(v), { message: 'Invalid numeric string' })",
);
const numberOrHex = new Code(
  'z.union([z.number(), hexString, numericString]).transform((n) => BigInt(n))',
);
const accountId = new Code(
  'z.union([hexString, z.string().regex(/^[0-9a-f]+$/).transform<`0x${string}`>((v) => `0x${v}`)]).transform((value) => ss58.encode({ data: value, ss58Format: 2112 }))',
  [new Identifier('* as ss58', '@chainflip/utils/ss58')],
);
const simpleEnum = new Code(
  '<U extends string, T extends readonly [U, ...U[]]>(values: T) => z.object({ __kind: z.enum(values) }).transform(({ __kind }) => __kind!)',
);

const shortChainToLongChain = {
  Arb: 'Arbitrum',
  Btc: 'Bitcoin',
  Eth: 'Ethereum',
  Dot: 'Polkadot',
  Sol: 'Solana',
} as const;

const chainEnumMember = (name: keyof typeof shortChainToLongChain, transform?: string) =>
  `z.object({ __kind: z.literal('${name}'), value: hexString }).transform(({ value }) => ({
  chain: '${shortChainToLongChain[name]}' as const,
  address: ${transform ?? 'value'},
}))`;

export default class CodeGenerator extends BaseCodeGenerator {
  private generateEncodedAddressEnum(def: EnumType): Identifier {
    this.registry.types.set('hexString', hexString);

    const dependencies: Identifier[] = [];

    const unionMembers = def.values.map((v) => {
      switch (v.name) {
        case 'Arb':
        case 'Eth':
          return chainEnumMember(v.name);
        case 'Dot':
          dependencies.push(new Identifier('* as ss58', '@chainflip/utils/ss58'));
          return chainEnumMember(v.name, `ss58.encode({ data: value, ss58Format: 0 })`);
        case 'Btc':
          return chainEnumMember(v.name, `Buffer.from(value.slice(2), 'hex').toString('utf8')`);
        case 'Sol':
          dependencies.push(
            new Identifier('* as base58', '@chainflip/utils/base58'),
            new Identifier('hexToBytes', '@chainflip/utils/bytes'),
          );
          return chainEnumMember(v.name, `base58.encode(hexToBytes(value))`);
        default:
          throw new Error(`unknown chain: "${v.name}"`);
      }
    });

    const generated = `z.union([${unionMembers.join(',')}])`;

    this.registry.types.set(def.name, new Code(generated, dependencies));

    return new Identifier(def.name);
  }

  protected override generatePrimitive(def: PrimitiveType): CodegenResult {
    switch (def.name) {
      case 'i8':
      case 'u8':
      case 'i16':
      case 'u16':
      case 'i32':
      case 'u32':
      case 'Percent':
      case 'Permill':
        return new Code('z.number()');
      case 'u64':
      case 'u128':
      case 'U256':
        this.registry.types.set('numericString', numericString);
        this.registry.types.set('hexString', hexString);
        this.registry.types.set('numberOrHex', numberOrHex);
        return new Identifier('numberOrHex');
      case 'AccountId32':
        this.registry.types.set('hexString', hexString);
        this.registry.types.set('accountId', accountId);
        return new Identifier('accountId');
      case 'H160':
      case 'H256':
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      case 'Call':
        // we would need to parse the calls as well
        return new Code('z.unknown()');
      case 'Bytes':
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      case 'bool':
        return new Code('z.boolean()');
      case 'null':
        return new Code('z.null()');
    }

    throw new Error(`Unsupported primitive: ${def.name}`);
  }

  protected override generateEnum(def: EnumType): Identifier {
    if (this.registry.types.has(def.name)) return new Identifier(def.name);

    if (def.name === 'cfChainsAddressEncodedAddress') {
      return this.generateEncodedAddressEnum(def);
    }

    // "isSimple" means that none of the variants have any associated data, e.g.
    // `enum Foo { A, B, C }` and not `enum Foo { A(u32), B { field: u32 }, C }`
    const isSimple = def.values.every((v) => isNull(v.value));
    // sometimes there will be empty placeholders, we can ignore these
    const values = def.values.filter((n) => !n.name.startsWith('__Unused'));

    let generated: string;
    const dependencies: CodegenResult[] = [];

    if (isSimple) {
      this.registry.types.set('simpleEnum', simpleEnum);
      generated = `simpleEnum([${values.map((v) => `'${v.name}'`).join(', ')}])`;
    } else {
      const members = values.map((v) => {
        if (isNull(v.value)) {
          return `z.object({ __kind: z.literal('${v.name}') })`;
        }

        // if the struct has no name, it is not actually a struct but an enum variant w/ named fields
        if (v.value.type === 'struct' && v.value.name === undefined) {
          const value = this.generateStruct({
            ...v.value,
            additionalFields: {
              __kind: `z.literal('${v.name}')`,
            },
          });

          dependencies.push(value);

          return value.toString();
        }

        const value = this.generateResolvedType(v.value);

        dependencies.push(value);

        return `z.object({ __kind: z.literal('${v.name}'), value: ${value.toString()} })`;
      });

      if (members.length === 1) {
        [generated] = members;
      } else {
        generated = `z.union([${members.join(', ')}])`;
      }
    }

    this.registry.types.set(def.name, new Code(generated, dependencies));

    return new Identifier(def.name);
  }

  protected override generateStruct(def: StructType): CodegenResult {
    const structIdent = def.name && nameToIdentifier(def.name);

    if (structIdent && this.registry.types.has(structIdent)) {
      return new Identifier(structIdent);
    }

    const dependencies: CodegenResult[] = [];

    const code = new Code(
      `z.object({ ${Object.entries(def.additionalFields ?? {})
        .map(([key, value]) => `${key}: ${value}`)
        .concat(
          Object.entries(def.fields)
            .filter(([, value]) => !isNull(value))
            .map(([key, value]) => {
              const resolvedType = this.generateResolvedType(value);
              dependencies.push(resolvedType);

              if (resolvedType instanceof Identifier) {
                const identifier = resolvedType.toString();

                if (identifier === key) return key;

                return `${key}: ${identifier}`;
              }

              return `${key}: ${resolvedType.toString()}`;
            }),
        )
        .join(', ')} })`,
      dependencies,
    );

    if (!structIdent) return code;

    this.registry.types.set(structIdent, code);

    return new Identifier(structIdent);
  }

  protected override generateArray(def: ArrayType): CodegenResult {
    if (def.length) {
      if (isPrimitiveType(def.value) && def.value.name === 'u8') {
        this.registry.types.set('hexString', hexString);
        return new Identifier('hexString');
      }

      const resolvedType = this.generateResolvedType(def.value);

      const dependencies = Array.from({ length: def.length }, () => resolvedType);

      return new Code(`z.tuple([${dependencies.join(', ')}])`, [resolvedType]);
    }

    const resolvedType = this.generateResolvedType(def.value);
    const dependencies =
      resolvedType instanceof Identifier ? [resolvedType] : [...resolvedType.dependencies];
    return new Code(`z.array(${resolvedType.toString()})`, dependencies);
  }

  protected override generateTuple(def: TupleType): Code {
    const dependencies = def.values.map((t) => {
      const resolvedType = this.generateResolvedType(t);
      return resolvedType;
    });

    return new Code(`z.tuple([${dependencies.join(', ')}])`, dependencies);
  }

  protected override generateOption(def: OptionType): Code {
    const resolvedType = this.generateResolvedType(def.value);
    return new Code(`${resolvedType.toString()}.nullish()`, [resolvedType]);
  }

  protected override generateRange(def: RangeType): Code {
    const resolvedType = this.generateResolvedType(def.value);
    return new Code(
      `z.object({ start: ${resolvedType.toString()}, end: ${resolvedType.toString()} })`,
      [resolvedType],
    );
  }

  protected override generateMap(def: MapType): Code {
    const keyType = this.generateResolvedType(def.key);
    const valueType = this.generateResolvedType(def.value);
    return new Code(`z.array(z.tuple([${keyType.toString()}, ${valueType.toString()}]))`, [
      keyType,
      valueType,
    ]);
  }

  protected override getParserName(palletName: string, itemName: string): string {
    return `${palletName}.${itemName}`;
  }

  protected override getName(palletName: string, itemName: string): string {
    return `${palletName}::${itemName}`
      .replace(/(?:::|_)(.)/g, (_, c: string) => c.toUpperCase())
      .replace(/^./, (c) => c.toLowerCase());
  }

  protected override getGlobalImports(): string[] {
    return ["import { z } from 'zod';"];
  }
}
