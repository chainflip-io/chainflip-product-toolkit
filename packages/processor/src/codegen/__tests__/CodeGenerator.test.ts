import * as fs from 'fs/promises';
import { describe, expect, it, vi } from 'vitest';
import CodeGenerator from '../CodeGenerator';

vi.mock('fs/promises');

const primitives = [
  'i8',
  'u8',
  'i16',
  'u16',
  'i32',
  'u32',
  'Percent',
  'Permill',
  'u64',
  'u128',
  'U256',
  'AccountId32',
  'H160',
  'H256',
  'Call',
  'Bytes',
  'bool',
  'null',
] as const;

describe(CodeGenerator, () => {
  it('generates a lot of stuff', async () => {
    const modules = [
      ...new CodeGenerator().generate({
        Primitives: Object.fromEntries(
          primitives.map((name) => [name, { type: 'primitive', name }]),
        ),
        Structs: {
          EventOne: {
            type: 'struct',
            fields: {
              ...Object.fromEntries(primitives.map((name) => [name, { type: 'primitive', name }])),
              ...Object.fromEntries(
                primitives.map((name) => [
                  `${name}Nullable`,
                  { type: 'option', value: { type: 'primitive', name } },
                ]),
              ),
              u32Array: {
                type: 'array',
                value: {
                  type: 'primitive',
                  name: 'u32',
                },
              },
              u32ArrayWithLength: {
                type: 'array',
                length: 3,
                value: {
                  type: 'primitive',
                  name: 'u32',
                },
              },
              bytes: {
                type: 'array',
                length: 32,
                value: {
                  type: 'primitive',
                  name: 'u8',
                },
              },
              tuple: {
                type: 'tuple',
                values: [
                  { type: 'primitive', name: 'u32' },
                  { type: 'primitive', name: 'bool' },
                  { type: 'primitive', name: 'AccountId32' },
                ],
              },
              range: {
                type: 'range',
                value: { type: 'primitive', name: 'u128' },
              },
              map: {
                type: 'map',
                key: {
                  type: 'tuple',
                  values: [
                    { type: 'primitive', name: 'AccountId32' },
                    { type: 'primitive', name: 'u32' },
                  ],
                },
                value: {
                  type: 'struct',
                  name: 'mapValue',
                  fields: {
                    value: { type: 'primitive', name: 'u32' },
                    moneySymbol: {
                      type: 'enum',
                      name: 'moneySymbol',
                      values: [{ name: 'Flip', value: { type: 'primitive', name: 'null' } }],
                    },
                  },
                },
              },
              flattenedEnumThing: {
                type: 'enum',
                name: 'enumWithFlattenedField',
                values: [
                  {
                    name: 'One',
                    value: {
                      type: 'struct',
                      fields: { field: { type: 'primitive', name: 'u32' } },
                    },
                  },
                ],
              },
              arrayOfIdentifiers: {
                type: 'array',
                value: {
                  type: 'enum',
                  name: 'moneySymbol',
                  values: [{ name: 'Flip', value: { type: 'primitive', name: 'null' } }],
                },
              },
            },
          },
        },
        Enums: {
          EventOne: {
            type: 'enum',
            name: 'myEnum',
            values: [
              ...primitives.map((name) => ({ name, value: { type: 'primitive' as const, name } })),
            ],
          },
        },
      }),
    ];

    for (const module of modules) {
      expect(await module.toFormattedString()).toMatchSnapshot(module['name']);
    }
  });

  it('generates nothing for an empty spec', () => {
    expect([...new CodeGenerator().generate({})]).toEqual([]);
  });

  it('filters on selected events', async () => {
    const modules = [
      ...new CodeGenerator({ trackedEvents: new Set(['Enums.EventOne']) }).generate({
        Primitives: Object.fromEntries(
          primitives.map((name) => [name, { type: 'primitive', name }]),
        ),
        Enums: {
          EventOne: {
            type: 'enum',
            name: 'myEnum',
            values: [
              { name: 'Flip', value: { type: 'primitive', name: 'null' } },
              { name: 'Usdc', value: { type: 'primitive', name: 'null' } },
              { name: 'Eth', value: { type: 'primitive', name: 'null' } },
            ],
          },
        },
      }),
    ];

    for (const module of modules) {
      expect(await module.toFormattedString()).toMatchSnapshot(module['name']);
    }
  });

  it('throws if it does not find the expected event', () => {
    vi.spyOn(console, 'error').mockImplementation(() => {});
    const it = new CodeGenerator({ trackedEvents: new Set(['Enums.EventOne']) }).generate({
      Primitives: Object.fromEntries(primitives.map((name) => [name, { type: 'primitive', name }])),
    });

    expect(() => it.next()).toThrowErrorMatchingInlineSnapshot(
      `[Error: Not all events were generated]`,
    );
    expect(console.error).toHaveBeenCalled();
  });

  it('throws on unhandled types', () => {
    vi.spyOn(console, 'error').mockImplementation(() => {});

    expect(() => [
      ...new CodeGenerator().generate({
        // @ts-expect-error -- that's the point
        Primitives: Object.fromEntries(
          primitives.map((name) => [name, { type: 'some new type', name }]),
        ),
      }),
    ]).toThrowErrorMatchingInlineSnapshot(`[Error: Unsupported type: some new type]`);
    expect(console.error).toHaveBeenCalled();
  });

  it('generates a named struct once', async () => {
    const modules = [
      ...new CodeGenerator().generate({
        PalletOne: {
          EventOne: {
            type: 'struct',
            name: 'reusedStruct',
            fields: {
              field: { type: 'primitive', name: 'u32' },
            },
          },
          AnotherEvent: {
            type: 'struct',
            name: 'reusedStruct',
            fields: {
              field: { type: 'primitive', name: 'u32' },
            },
          },
        },
      }),
    ];

    for (const module of modules) {
      expect(await module.toFormattedString()).toMatchSnapshot(module['name']);
    }
  });

  it('throws on unsupported primitives', () => {
    vi.spyOn(console, 'error').mockImplementation(() => {});
    expect(() => [
      ...new CodeGenerator().generate({
        Primitives: {
          Event: { type: 'primitive', name: 'unknown' },
        },
      }),
    ]).toThrowErrorMatchingInlineSnapshot(`[Error: Unsupported primitive: unknown]`);
    expect(console.error).toHaveBeenCalled();
  });

  it('writes a formateted module to disk', async () => {
    const spy = vi.spyOn(fs, 'writeFile');

    const modules = [
      ...new CodeGenerator().generate({
        PalletOne: {
          EventOne: {
            type: 'struct',
            name: 'reusedStruct',
            fields: {
              field: { type: 'primitive', name: 'u32' },
            },
          },
        },
      }),
    ];

    for (const module of modules) {
      await module.writeFile('topLevelDir');
    }

    expect(spy.mock.calls).toMatchInlineSnapshot(`
      [
        [
          "topLevelDir/palletOne/eventOne.ts",
          "import { z } from 'zod';
      import { reusedStruct } from '../common';

      export const palletOneEventOne = reusedStruct;
      ",
          "utf8",
        ],
        [
          "topLevelDir/common.ts",
          "import { z } from 'zod';

      export const reusedStruct = z.object({ field: z.number() });
      ",
          "utf8",
        ],
      ]
    `);
  });
});
