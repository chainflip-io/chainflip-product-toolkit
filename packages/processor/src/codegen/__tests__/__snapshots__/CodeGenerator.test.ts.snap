// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeGenerator > filters on selected events > EventOne 1`] = `
"import { z } from 'zod';
import { myEnum } from '../common';

export const enumsEventOne = myEnum;
"
`;

exports[`CodeGenerator > filters on selected events > common 1`] = `
"import { z } from 'zod';

export const simpleEnum = <U extends string, T extends readonly [U, ...U[]]>(values: T) =>
  z.object({ __kind: z.enum(values) }).transform(({ __kind }) => __kind!);

export const myEnum = simpleEnum(['Flip', 'Usdc', 'Eth']);
"
`;

exports[`CodeGenerator > generates a lot of stuff > AccountId32 1`] = `
"import { z } from 'zod';
import { accountId } from '../common';

export const primitivesAccountId32 = accountId;
"
`;

exports[`CodeGenerator > generates a lot of stuff > Bytes 1`] = `
"import { z } from 'zod';
import { hexString } from '../common';

export const primitivesBytes = hexString;
"
`;

exports[`CodeGenerator > generates a lot of stuff > Call 1`] = `
"import { z } from 'zod';

export const primitivesCall = z.unknown();
"
`;

exports[`CodeGenerator > generates a lot of stuff > EventOne 1`] = `
"import { z } from 'zod';
import {
  accountId,
  enumWithFlattenedField,
  hexString,
  mapValue,
  moneySymbol,
  numberOrHex,
} from '../common';

export const structsEventOne = z.object({
  i8: z.number(),
  u8: z.number(),
  i16: z.number(),
  u16: z.number(),
  i32: z.number(),
  u32: z.number(),
  Percent: z.number(),
  Permill: z.number(),
  u64: numberOrHex,
  u128: numberOrHex,
  U256: numberOrHex,
  AccountId32: accountId,
  H160: hexString,
  H256: hexString,
  Call: z.unknown(),
  Bytes: hexString,
  bool: z.boolean(),
  i8Nullable: z.number().nullish(),
  u8Nullable: z.number().nullish(),
  i16Nullable: z.number().nullish(),
  u16Nullable: z.number().nullish(),
  i32Nullable: z.number().nullish(),
  u32Nullable: z.number().nullish(),
  PercentNullable: z.number().nullish(),
  PermillNullable: z.number().nullish(),
  u64Nullable: numberOrHex.nullish(),
  u128Nullable: numberOrHex.nullish(),
  U256Nullable: numberOrHex.nullish(),
  AccountId32Nullable: accountId.nullish(),
  H160Nullable: hexString.nullish(),
  H256Nullable: hexString.nullish(),
  CallNullable: z.unknown().nullish(),
  BytesNullable: hexString.nullish(),
  boolNullable: z.boolean().nullish(),
  nullNullable: z.null().nullish(),
  u32Array: z.array(z.number()),
  u32ArrayWithLength: z.tuple([z.number(), z.number(), z.number()]),
  bytes: hexString,
  tuple: z.tuple([z.number(), z.boolean(), accountId]),
  range: z.object({ start: numberOrHex, end: numberOrHex }),
  map: z.array(z.tuple([z.tuple([accountId, z.number()]), mapValue])),
  flattenedEnumThing: enumWithFlattenedField,
  arrayOfIdentifiers: z.array(moneySymbol),
});
"
`;

exports[`CodeGenerator > generates a lot of stuff > EventOne 2`] = `
"import { z } from 'zod';
import { myEnum } from '../common';

export const enumsEventOne = myEnum;
"
`;

exports[`CodeGenerator > generates a lot of stuff > H160 1`] = `
"import { z } from 'zod';
import { hexString } from '../common';

export const primitivesH160 = hexString;
"
`;

exports[`CodeGenerator > generates a lot of stuff > H256 1`] = `
"import { z } from 'zod';
import { hexString } from '../common';

export const primitivesH256 = hexString;
"
`;

exports[`CodeGenerator > generates a lot of stuff > Percent 1`] = `
"import { z } from 'zod';

export const primitivesPercent = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > Permill 1`] = `
"import { z } from 'zod';

export const primitivesPermill = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > U256 1`] = `
"import { z } from 'zod';
import { numberOrHex } from '../common';

export const primitivesU256 = numberOrHex;
"
`;

exports[`CodeGenerator > generates a lot of stuff > bool 1`] = `
"import { z } from 'zod';

export const primitivesBool = z.boolean();
"
`;

exports[`CodeGenerator > generates a lot of stuff > common 1`] = `
"import { z } from 'zod';
import * as ss58 from '@chainflip/utils/ss58';

export const numericString = z
  .string()
  .refine((v) => /^\\d+$/.test(v), { message: 'Invalid numeric string' });

export const hexString = z
  .string()
  .refine((v): v is \`0x\${string}\` => /^0x[\\da-f]*$/i.test(v), { message: 'Invalid hex string' });

export const numberOrHex = z
  .union([z.number(), hexString, numericString])
  .transform((n) => BigInt(n));

export const accountId = z
  .union([
    hexString,
    z
      .string()
      .regex(/^[0-9a-f]+$/)
      .transform<\`0x\${string}\`>((v) => \`0x\${v}\`),
  ])
  .transform((value) => ss58.encode({ data: value, ss58Format: 2112 }));

export const simpleEnum = <U extends string, T extends readonly [U, ...U[]]>(values: T) =>
  z.object({ __kind: z.enum(values) }).transform(({ __kind }) => __kind!);

export const moneySymbol = simpleEnum(['Flip']);

export const mapValue = z.object({ value: z.number(), moneySymbol });

export const enumWithFlattenedField = z.object({ __kind: z.literal('One'), field: z.number() });

export const myEnum = z.union([
  z.object({ __kind: z.literal('i8'), value: z.number() }),
  z.object({ __kind: z.literal('u8'), value: z.number() }),
  z.object({ __kind: z.literal('i16'), value: z.number() }),
  z.object({ __kind: z.literal('u16'), value: z.number() }),
  z.object({ __kind: z.literal('i32'), value: z.number() }),
  z.object({ __kind: z.literal('u32'), value: z.number() }),
  z.object({ __kind: z.literal('Percent'), value: z.number() }),
  z.object({ __kind: z.literal('Permill'), value: z.number() }),
  z.object({ __kind: z.literal('u64'), value: numberOrHex }),
  z.object({ __kind: z.literal('u128'), value: numberOrHex }),
  z.object({ __kind: z.literal('U256'), value: numberOrHex }),
  z.object({ __kind: z.literal('AccountId32'), value: accountId }),
  z.object({ __kind: z.literal('H160'), value: hexString }),
  z.object({ __kind: z.literal('H256'), value: hexString }),
  z.object({ __kind: z.literal('Call'), value: z.unknown() }),
  z.object({ __kind: z.literal('Bytes'), value: hexString }),
  z.object({ __kind: z.literal('bool'), value: z.boolean() }),
  z.object({ __kind: z.literal('null') }),
]);
"
`;

exports[`CodeGenerator > generates a lot of stuff > i8 1`] = `
"import { z } from 'zod';

export const primitivesI8 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > i16 1`] = `
"import { z } from 'zod';

export const primitivesI16 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > i32 1`] = `
"import { z } from 'zod';

export const primitivesI32 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > null 1`] = `
"import { z } from 'zod';

export const primitivesNull = z.null();
"
`;

exports[`CodeGenerator > generates a lot of stuff > u8 1`] = `
"import { z } from 'zod';

export const primitivesU8 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > u16 1`] = `
"import { z } from 'zod';

export const primitivesU16 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > u32 1`] = `
"import { z } from 'zod';

export const primitivesU32 = z.number();
"
`;

exports[`CodeGenerator > generates a lot of stuff > u64 1`] = `
"import { z } from 'zod';
import { numberOrHex } from '../common';

export const primitivesU64 = numberOrHex;
"
`;

exports[`CodeGenerator > generates a lot of stuff > u128 1`] = `
"import { z } from 'zod';
import { numberOrHex } from '../common';

export const primitivesU128 = numberOrHex;
"
`;

exports[`CodeGenerator > generates a named struct once > AnotherEvent 1`] = `
"import { z } from 'zod';
import { reusedStruct } from '../common';

export const palletOneAnotherEvent = reusedStruct;
"
`;

exports[`CodeGenerator > generates a named struct once > EventOne 1`] = `
"import { z } from 'zod';
import { reusedStruct } from '../common';

export const palletOneEventOne = reusedStruct;
"
`;

exports[`CodeGenerator > generates a named struct once > common 1`] = `
"import { z } from 'zod';

export const reusedStruct = z.object({ field: z.number() });
"
`;

exports[`CodeGenerator > generates the encoded address enum > EventOne 1`] = `
"import { z } from 'zod';
import { cfChainsAddressEncodedAddress } from '../common';

export const palletOneEventOne = cfChainsAddressEncodedAddress;
"
`;

exports[`CodeGenerator > generates the encoded address enum > common 1`] = `
"import { z } from 'zod';
import * as ss58 from '@chainflip/utils/ss58';
import * as base58 from '@chainflip/utils/base58';
import { hexToBytes } from '@chainflip/utils/bytes';

export const hexString = z
  .string()
  .refine((v): v is \`0x\${string}\` => /^0x[\\da-f]*$/i.test(v), { message: 'Invalid hex string' });

export const cfChainsAddressEncodedAddress = z.union([
  z.object({ __kind: z.literal('Eth'), value: hexString }).transform(({ value }) => ({
    chain: 'Ethereum' as const,
    address: value,
  })),
  z.object({ __kind: z.literal('Dot'), value: hexString }).transform(({ value }) => ({
    chain: 'Polkadot' as const,
    address: ss58.encode({ data: value, ss58Format: 0 }),
  })),
  z.object({ __kind: z.literal('Sol'), value: hexString }).transform(({ value }) => ({
    chain: 'Solana' as const,
    address: base58.encode(hexToBytes(value)),
  })),
  z.object({ __kind: z.literal('Btc'), value: hexString }).transform(({ value }) => ({
    chain: 'Bitcoin' as const,
    address: Buffer.from(value.slice(2), 'hex').toString('utf8'),
  })),
  z.object({ __kind: z.literal('Arb'), value: hexString }).transform(({ value }) => ({
    chain: 'Arbitrum' as const,
    address: value,
  })),
]);
"
`;
